### PR TITLE
Add support for bundling source code in DownloaderCommand

### PR DESCRIPTION
**Pull request by KorSin**

### CODE REVIEW

File: cli/src/main/kotlin/commands/DownloaderCommand.kt [hunk 1]
```diff
@@ -118,7 +120,13 @@ class DownloaderCommand : CliktCommand(name = "download", help = "Fetch source c
     private val archive by option(
         "--archive",
         help = "Archive the downloaded source code as ZIP files to the output directory. Is ignored if " +
-                "'--project-url' is also specified."
+                "one of 'archive-all' and '--project-url' is also specified."
```
Comment (sschuberth) [inline cli/src/main/kotlin/commands/DownloaderCommand.kt:123]: s/archive-all/--archive-all/
s/and/or/

File: cli/src/main/kotlin/commands/DownloaderCommand.kt [hunk 2]
```diff
@@ -118,7 +120,13 @@ class DownloaderCommand : CliktCommand(name = "download", help = "Fetch source c
...
+    ).flag()
+
+    private val archiveAll by option(
+        "--archive-all",
+        help = "Archive the all the downloaded source code as a single zip file to the output directory. Is ignored " +
```
Comment (sschuberth) [inline cli/src/main/kotlin/commands/DownloaderCommand.kt:128]: s/zip/ZIP/

File: cli/src/main/kotlin/commands/DownloaderCommand.kt [hunk 3]
```diff
@@ -118,7 +120,13 @@ class DownloaderCommand : CliktCommand(name = "download", help = "Fetch source c
...
+                "if '--project-url' is also specified."
```
Comment (sschuberth) [inline cli/src/main/kotlin/commands/DownloaderCommand.kt:129]: "if one of '--archive' or '--project-url' is also specified."

File: cli/src/main/kotlin/commands/DownloaderCommand.kt [hunk 4]
```diff
@@ -247,4 +257,15 @@ class DownloaderCommand : CliktCommand(name = "download", help = "Fetch source c
...
+
+    private fun archiveAll() {
+        val zipFile = createTempFile(ORT_NAME, ".zip").toFile()
+        zipFile.deleteOnExit()
+
+        outputDir.packZip(zipFile, overwrite = true)
+
+        outputDir.deleteRecursively()
+
+        zipFile.copyTo(outputDir.resolve("archive.zip"))
```
Comment (sschuberth) [inline cli/src/main/kotlin/commands/DownloaderCommand.kt:269]: Why don't you create the ZIP in the final location in the first place?

File: cli/src/main/kotlin/commands/DownloaderCommand.kt [hunk 5]
```diff
@@ -247,4 +257,15 @@ class DownloaderCommand : CliktCommand(name = "download", help = "Fetch source c
             outputDir.resolve(relativePath.toString()).safeDeleteRecursively()
         }
     }
+
+    private fun archiveAll() {
+        val zipFile = createTempFile(ORT_NAME, ".zip").toFile()
+        zipFile.deleteOnExit()
+
+        outputDir.packZip(zipFile, overwrite = true)
```
Comment (sschuberth) [inline cli/src/main/kotlin/commands/DownloaderCommand.kt:265]: We shouldn't unconditionally pass `overwrite = true` here as we also don't to that in `archive()`.

File: cli/src/main/kotlin/commands/ScannerCommand.kt [hunk 6]
```diff
@@ -118,6 +121,11 @@ class ScannerCommand : CliktCommand(name = "scan", help = "Run existing copyrigh
         help = "Do not scan excluded projects or packages. Works only with the '--ort-file' parameter."
     ).flag()
 
+    private val archiveAll by option(
```
Comment (sschuberth) [inline cli/src/main/kotlin/commands/ScannerCommand.kt:124]: Let's move this commit to a separate PR as I believe it will trigger longer discussions.

**Review by sschuberth — CHANGES_REQUESTED** [review]

File: cli/src/main/kotlin/commands/DownloaderCommand.kt (hunk 3 above)
Comment (KorSin) [reply cli/src/main/kotlin/commands/DownloaderCommand.kt:129]: In the current logic, '--archive-all' takes the precedence over '--archive'

File: cli/src/main/kotlin/commands/DownloaderCommand.kt (hunk 4 above)
Comment (KorSin) [reply cli/src/main/kotlin/commands/DownloaderCommand.kt:269]: If you create the ZIP in the download directory and then try to pack everything in the directory, an infinite loop is created. It tries to pack itself and so on. However, I think the `pack` method allows an additional filter method, so this could be done with that.

File: cli/src/main/kotlin/commands/DownloaderCommand.kt (hunk 5 above)
Comment (KorSin) [reply cli/src/main/kotlin/commands/DownloaderCommand.kt:265]: I guess you are right, but omitting the `overwrite = true` parameter creates an error. This is due to the temp file that has already been created when the `packZip` method is called. Apparently, even this files may not exist if no overwriting is allowed.

File: cli/src/main/kotlin/commands/DownloaderCommand.kt (hunk 4 above)
Comment (KorSin) [reply cli/src/main/kotlin/commands/DownloaderCommand.kt]: I decided to use the filter method, it reduces the code a lot.

File: cli/src/main/kotlin/commands/DownloaderCommand.kt (hunk 5 above)
Comment (KorSin) [reply cli/src/main/kotlin/commands/DownloaderCommand.kt]: Not using a temp file, makes this unnecessary.

File: cli/src/main/kotlin/commands/DownloaderCommand.kt (hunk 3 above)
Comment (sschuberth) [reply cli/src/main/kotlin/commands/DownloaderCommand.kt:127]: Then I believe this should explicitly be mentioned also in this help text by inserting "Takes precedence over '--archive'." after the first sentence.

File: cli/src/main/kotlin/commands/DownloaderCommand.kt [hunk 7]
```diff
@@ -118,7 +118,13 @@ class DownloaderCommand : CliktCommand(name = "download", help = "Fetch source c
     private val archive by option(
         "--archive",
         help = "Archive the downloaded source code as ZIP files to the output directory. Is ignored if " +
-                "'--project-url' is also specified."
+                "one of '--archive-all' or '--project-url' is also specified."
+    ).flag()
+
+    private val archiveAll by option(
+        "--archive-all",
+        help = "Archive the all the downloaded source code as a single ZIP file to the output directory. Is ignored " +
```
Comment (sschuberth) [inline cli/src/main/kotlin/commands/DownloaderCommand.kt:126]: s/the all/all the/

File: cli/src/main/kotlin/commands/DownloaderCommand.kt [hunk 8]
```diff
@@ -247,4 +255,12 @@ class DownloaderCommand : CliktCommand(name = "download", help = "Fetch source c
             outputDir.resolve(relativePath.toString()).safeDeleteRecursively()
         }
     }
+
+    private fun archiveAll() {
+        val archiveFile = outputDir.resolve("archive.zip")
```
Comment (sschuberth) [inline cli/src/main/kotlin/commands/DownloaderCommand.kt:260]: Let call this `zipFile` analog to in `archive()`.

File: cli/src/main/kotlin/commands/DownloaderCommand.kt [hunk 9]
```diff
@@ -247,4 +255,12 @@ class DownloaderCommand : CliktCommand(name = "download", help = "Fetch source c
...
+
```
Comment (sschuberth) [inline cli/src/main/kotlin/commands/DownloaderCommand.kt:261]: Please add in `info` log analog to in line 240 ff.

File: cli/src/main/kotlin/commands/DownloaderCommand.kt [hunk 10]
```diff
@@ -247,4 +255,12 @@ class DownloaderCommand : CliktCommand(name = "download", help = "Fetch source c
...
+        outputDir.packZip(archiveFile) { path ->
+            path != archiveFile.toPath()
```
Comment (sschuberth) [inline cli/src/main/kotlin/commands/DownloaderCommand.kt:263]: As a small optimization, you could move out the path conversion so it's not called on every lambda invocation by adding

    val zipPath = zipFile.toPath()

to line 261.

**Review by sschuberth — CHANGES_REQUESTED** [review]

**Review by sschuberth — APPROVED** [review]
